gorm/transaction.go

```diff
@@ -57,15 +81,29 @@ func (t *Transaction) AddAfterCommitHook(hooks ...func(context.Context)) {
 	t.afterCommitHook = append(t.afterCommitHook, hooks...)
 }
 
-// BeginFromContext will extract transaction wrapper from context and start new transaction.
+// GetReadOnlyDBInstance returns the read only database instance stored in the ctx
+func GetReadOnlyDBInstance(ctx context.Context) (*gorm.DB, error) {
+	dbRO, ok := ctx.Value(roDBKey).(*gorm.DB)
+	if !ok {
+		return nil, ErrNoReadOnlyDB
+	}
+	return dbRO, nil
+}
+
+// BeginFromContext will extract transaction wrapper from context and start new transaction if transaction is not set to read only otherwise it will return read only database instance
```

Comment:
Not using an actual transaction for read-only operations introduces two problems:
1. No support for isolation level,
2. Performance (see `Statements are executed more quickly in a transaction block, because transaction start/commit requires significant CPU and disk activity.` in https://www.postgresql.org/docs/15/sql-begin.html).

Reply:
This point is not addressed. The code still returns the "database" and not a transaction.